lib/puma/server.rb

```diff
@@ -681,6 +705,30 @@ def handle_request(req, lines)
           status, headers, res_body = lowlevel_error(e, env)
         end
 
+        if @on_after_rack
+          is_async = false
+
+          @on_after_rack.reverse_each do |hook|
+            stream_client = hook.call(env, headers, req.io)
+
+            if stream_client && stream_client.stream?
+              if is_async
+                raise "Only one #on_after_rack hook should take over"
```

Comment:
I think we can avoid this by just giving the user some notion of an "insertion order" - insert this plugin at this point the stack, then this plugin, etc, and the first plugin that is_async gets to take over.